File: baekjoon/algo_1937.py
N = int(input())
bambus = []
dp = [ [ 0 for _ in range(N)] for _ in range(N)]
direction = [(0,1), (0,-1), (1, 0), (-1, 0)]

ans = 0
count = 0
for i in range(N):
    bambus.append([ int(x) for x in input().split() ])

def dfs(y, x):
    global dp
    global ans
    
    value = 0
    for d in direction:
        nx = x + d[0]
        ny = y + d[1]
                
        if nx < 0 or nx >= N or ny < 0 or ny >= N:
            continue
        
        if bambus[ny][nx] > bambus[y][x]:
            if dp[ny][nx] == 0:
                dfs(nx, ny)
                
            if value < dp[ny][nx]:
                value = dp[ny][nx]
    
    dp[y][x] = value + 1
    
    if ans < dp[y][x]:
        ans = dp[y][x]
            
for i in range(N):
    for j in range(N):
        if dp[i][j] == 0:
            dfs(i, j)
print(ans)

File: baekjoon/algo_1937.cpp
#include<stdio.h>

int bambus[501][501];
int dp[501][501];

int dx[4] = { 0, 0 ,1 ,-1};
int dy[4] = { -1, 1, 0 ,0};

int N;
int ans;

void dfs(int cy, int cx)
{
    int local_m = 0;  
    
    for(int i = 0 ; i< 4; i++)
    {
        int nx = cx + dx[i];
        int ny = cy + dy[i];
        
        if(nx < 0 || nx >= N || ny < 0 || ny >= N)
            continue;
        
        if(bambus[ny][nx] > bambus[cy][cx])
        {
            if(dp[ny][nx] == 0)
                dfs(ny, nx);
                
            if(local_m < dp[ny][nx])
                local_m = dp[ny][nx];
        }
    }
    
    dp[cy][cx] = local_m + 1;
        
    if(ans < dp[cy][cx])
        ans = dp[cy][cx];
    
}

int main()
{
    scanf("%d", &N);
    
    for(int i =0 ; i< N; i++)
        for(int j=0; j < N; j++)
            scanf("%d", &bambus[i][j]);
            
    for(int i =0 ; i< N; i++)
        for(int j=0; j < N; j++)
            if(dp[i][j] == 0)
                dfs(i, j);
                
    for(int i =0 ; i< N; i++)
    {
        for(int j=0; j < N; j++)
            printf("%d", dp[i][j]);
        printf("\n");    
    }
                
    printf("%d", ans);
    
    return 0;
}